File: ts/src/validate.ts
// 201: a second agent judges the first.
//
//   npm run validate -- "task"
//
// Runs the loop, hands the trace and criteria.md to a reviewer, gets a structured
// verdict. Rejected runs go back to the loop with the feedback, up to MAX_ATTEMPTS.
// After that a human decides. That is the point: you review verdicts, not runs.
import fs from "node:fs";
import path from "node:path";
import Anthropic from "@anthropic-ai/sdk";
import { zodOutputFormat } from "@anthropic-ai/sdk/helpers/zod";
import { z } from "zod";
import { MODEL, here } from "./env";
import * as loop from "./loop";

const CRITERIA = fs.readFileSync(path.join(here, "criteria.md"), "utf8");
const MAX_ATTEMPTS = 3;
const client = new Anthropic();

const Verdict = z.object({
  approved: z.boolean(),
  score: z.number().int(),
  feedback: z.string(),
});
export type Verdict = z.infer<typeof Verdict>;

export async function validate(trace: loop.Trace): Promise<Verdict> {
  const response = await client.messages.parse({
    model: MODEL,
    max_tokens: 16000,
    system: "You are a strict reviewer. Judge the run against the criteria. Never fix the work yourself.",
    messages: [{ role: "user", content: `# Criteria\n${CRITERIA}\n\n# Run\n${JSON.stringify(trace, null, 2)}` }],
    output_config: { format: zodOutputFormat(Verdict) },
  });
  if (!response.parsed_output) throw new Error(`reviewer returned no verdict (stop_reason=${response.stop_reason})`);
  return response.parsed_output;
}

export async function runValidated(task: string): Promise<loop.Trace> {
  let feedback = "";
  let trace!: loop.Trace;
  for (let attempt = 1; attempt <= MAX_ATTEMPTS; attempt++) {
    const prompt = feedback ? `${task}\n\nA reviewer rejected your last attempt. Their feedback:\n${feedback}` : task;
    trace = await loop.run(prompt);
    const verdict = await validate(trace);
    trace.verdict = { ...verdict, attempt };
    loop.save(trace);
    if (verdict.approved) return trace;
    feedback = verdict.feedback;
  }
  return trace;
}

if (process.argv[1]?.endsWith("validate.ts")) {
  const task = process.argv.slice(2).join(" ") || "What time is it? Remember it under last_run.";
  const t = await runValidated(task);
  console.log(t.output);
  console.error(`\nverdict: ${JSON.stringify(t.verdict, null, 2)}`);
}

File: python/agent/validate.py
"""201: a second agent judges the first.

    python -m agent.validate "task"

Runs the loop, hands the trace and criteria.md to a reviewer, gets a structured
verdict. Rejected runs go back to the loop with the feedback, up to MAX_ATTEMPTS.
After that a human decides. That is the point: you review verdicts, not runs.
"""
import json
import sys
from pathlib import Path

import anthropic
from pydantic import BaseModel

from agent import loop

CRITERIA = (Path(__file__).parent / "criteria.md").read_text()
MAX_ATTEMPTS = 3

client = anthropic.Anthropic()


class Verdict(BaseModel):
    approved: bool
    score: int
    feedback: str


def validate(trace: dict) -> Verdict:
    response = client.messages.parse(
        model=loop.MODEL,
        max_tokens=16000,
        system="You are a strict reviewer. Judge the run against the criteria. Never fix the work yourself.",
        messages=[
            {
                "role": "user",
                "content": f"# Criteria\n{CRITERIA}\n\n# Run\n{json.dumps(trace, indent=2, default=str)}",
            }
        ],
        output_format=Verdict,
    )
    return response.parsed_output


def run_validated(task: str) -> dict:
    feedback = ""
    for attempt in range(1, MAX_ATTEMPTS + 1):
        prompt = task if not feedback else f"{task}\n\nA reviewer rejected your last attempt. Their feedback:\n{feedback}"
        trace = loop.run(prompt)
        verdict = validate(trace)
        trace["verdict"] = verdict.model_dump() | {"attempt": attempt}
        loop.save(trace)
        if verdict.approved:
            return trace
        feedback = verdict.feedback
    return trace


if __name__ == "__main__":
    task = " ".join(sys.argv[1:]) or "What time is it? Remember it under last_run."
    t = run_validated(task)
    print(t["output"])
    print(f"\nverdict: {json.dumps(t['verdict'], indent=2)}", file=sys.stderr)
